10up/.github
# Update list of GPL-compatible licenses

### Description of the Change

**_This is going to be a bit of a long one, so buckle-up and bear with me as I take you through my assessment here..._**

As originally caught in the [Dependency Review check](https://github.com/10up/insert-special-characters/actions/runs/4005582567/jobs/6876110985) on https://github.com/10up/insert-special-characters/pull/173, three dependencies of Insert Special Characters were flagged as incompatible with the list in our [GPL-Compatible License Policy file](https://github.com/10up/.github/blob/trunk/.github/dependency-review-config.yml) (because they were all part of a newly included dependency 🤷🏼):

Here's the text of the failures:

```
  The following dependencies have incompatible licenses:
  package-lock.json » caniuse-lite@1.0.30001439 – License: CC-BY-4.0
  package-lock.json » mousetrap@1.6.5 – License: Apache-2.0 AND Apache-2.0 WITH LLVM-exception
  package-lock.json » tslib@2.4.1 – License: 0BSD
  Error: Dependency review detected incompatible licenses.
```

...and here's a screenshot of the failures (both pulled from the Dependency Review check linked above):

![Screenshot 2023-01-26 at 1 06 19 PM](https://user-images.githubusercontent.com/2818133/214958602-acf19b27-fcd3-4a0a-8319-3ed094517daa.png)

As it relates to [caniuse-lite](https://www.npmjs.com/package/caniuse-lite), [it uses the `CC-BY-4.0` license](https://github.com/browserslist/caniuse-lite/blob/6494e06c3ea33d49cfe62be59b6a0ef0e717bb16/package.json#L39) which is listed as [GPL-compatible on GNU.org](https://www.gnu.org/licenses/license-list.en.html#ccby).  On the [SPDX License List table](https://spdx.org/licenses/), the [`CC-BY-4.0` license](https://spdx.org/licenses/CC-BY-4.0.html) is listed as FSF Free/Libre but NOT OSI Approved.  While I would prefer to see that our dependencies are BOTH FSF and OSI approved, in the case of ensuring our projects can properly maintain their GPL licensing it's most important that we maintain the GPL-compatibility.

As it relates to [Mousetrap](https://www.npmjs.com/package/mousetrap), [it uses the `Apache-2.0 WITH LLVM-exception` license](https://github.com/ccampbell/mousetrap/blob/2f9a476ba6158ba69763e4fcf914966cc72ef433/package.json#L22) which utilizes the base `Apache-2.0` license that is listed as [GPL-compatible on GNU.org](https://www.gnu.org/licenses/license-list.en.html#apache2) and the [SPDX License List table](https://spdx.org/licenses/) shows [`Apache-2.0`](https://spdx.org/licenses/Apache-2.0.html) as both FSF Free/Libre AND [OSI Approved](https://opensource.org/licenses/Apache-2.0).  The [`LLVM Exception`](https://spdx.org/licenses/LLVM-exception.html) is noted on the [SPDX License Exceptions table](https://spdx.org/licenses/exceptions-index.html) and by including the `WITH` joiner to the base `Apache-2.0` license modifies that license by adding an exception for GPLv2 compatibility.

As it relates to [tslib](https://www.npmjs.com/package/tslib), [it uses the `0BSD` license](https://github.com/microsoft/tslib/blob/e388a23287bdea2c4f411d3ddedd1c80002f76a7/package.json#L6) which [unfortunately is not listed on GNU.org](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).  On the [SPDX License List table](https://spdx.org/licenses/), the [`0BSD` license](https://spdx.org/licenses/0BSD) is listed as [OSI Approved](https://opensource.org/licenses/0BSD) but is NOT FSF Free/Libre.  In further researching the [BSD Zero Clause License](https://choosealicense.com/licenses/0bsd/), it appears that it "goes further than the BSD 2-Clause license to allow you unlimited freedom with the software without requirements to include the copyright notice, license text, or disclaimer in either source or binary forms."  The [BSD 2-Clause / FreeBSD license is listed as GPL-compatible on GNU.org](https://www.gnu.org/licenses/license-list.en.html#FreeBSD) and the ["BSD Zero Clause License" itself is noted as GPL compatible on CodeDocs.org](https://codedocs.org/what-is/bsd-licenses).  [Finally you an see a grand chap such as @pento also weighed in on the `0BSD` license and similarly felt it was good to include as GPL-compatible for Gutenberg](https://github.com/WordPress/gutenberg/pull/22391#pullrequestreview-413208947) (not ironically [when `tslib` was approved for use in Gutenberg](https://github.com/WordPress/gutenberg/pull/22391#pullrequestreview-413208947)).

After some research, it appears that we can add the following to our list of GPL-compatible licenses:
- 0BSD
- Apache-2.0 WITH LLVM-exception
- CC-BY-4.0

Note that this PR also first alphabetized the order of licenses listed in the [dependency-review-config.yml file](https://github.com/10up/.github/blob/trunk/.github/dependency-review-config.yml) to make it easier to scan the list to find existing/non-existing licenses.

### How to test the Change

Check out the Insert Special Characters repo, update the dependency check to use the amended policy file here, run the updated check on the https://github.com/10up/insert-special-characters/pull/173 PR to see that the check now passes.

### Changelog Entry

> Added - `0BSD` to the list of GPL-compatible licenses in the Dependency Review License Policy.
> Added - `Apache-2.0 WITH LLVM-exception` to the list of GPL-compatible licenses in the Dependency Review License Policy.
> Added - `CC-BY-4.0` to the list of GPL-compatible licenses in the Dependency Review License Policy.

### Credits

Props @jeffpaul.

### Checklist:
